### ISSUE TITLE
Axis domain optional min/max support

### ISSUE THREAD

**Issue by sulemanof** [elastic/elastic-charts#111]
**Is your feature request related to a problem? Please describe.**
There is no way to pass a domain prop to the `Axis` component as an empty object or an object with partial values (only `max` or only `min` value). 

**Describe the solution you'd like**
I think it should support the next object structures:
- `const domain = { min: undefined || null, max: undefined || null}` and pretty the same `const domain = {}` - in that case it should behave as without the domain prop;
- `const domain = { min: 10 }` or `const domain = { max: 10 }` - support of partial domain values if a user wants to use only one of them

**Additional context**
In that case I would never think about if `min` or `max` is defined and it would simplify the development
![image](https://user-images.githubusercontent.com/31325372/54743210-9109cf80-4bd4-11e9-9f8a-a96c2ab87b7e.png)

- I suppose the charts should show a message that user pass invalid max-min range instead of throwing an error which crash an app
![image](https://user-images.githubusercontent.com/31325372/54743760-1e99ef00-4bd6-11e9-8550-65489b89d5dd.png)
![image](https://user-images.githubusercontent.com/31325372/54743777-31acbf00-4bd6-11e9-9164-59e360640c0f.png)

**Kibana Cross Issues**
n/a

**Checklist**

- [x] this request is checked against already exist requests
- ~[ ] every related Kibana issue is listed under `Kibana Cross Issues` list~
- ~[ ] `kibana cross issue` tag is associated to the issue if any kibana cross issue is present~

**Comment by markov00** [comment]
Making min and max optionals is a nice enhancement. @emmacunningham could you take a look into this?
@sulemanof For the error throw related question: we can have two way to handle this:
- let's the developer catch that error and notify the user with the appropriate message
- display a message instead of the chart with that error.

Can happen that the final user is not the one that have changed the domain min/max on the chart: think about the case where you are integrating the library and you always precompute the domain min and max to adjust them a bit, if you have an error on your code, we can display a default message but the user doesn't know how to deal with it, we should than add a way to integrate custom/translated messages for each error message available.

I think we need to discuss this a bit more to get a correct understanding on how to deal with errors

**Comment by markov00** [comment]
@sulemanof On the other way, we don't want to support null, undefined values or empty object checks. The typescript definition is rigid on the type of data to be passed and it already checks if the value is valid or not at compile time. 

We can change the `DomainRange` interface to something like this
```
// copied from stackoverflow :)
type RequireAtLeastOne<T, Keys extends keyof T = keyof T> =
    Pick<T, Exclude<keyof T, Keys>>
    & {
        [K in Keys]-?: Required<Pick<T, K>>
    }[Keys]

type DomainRange = RequireAtLeastOne<Record<'min' | 'max', number>, 'min' | 'max'>;

const a: DomainRange = {}; // not valid
const b: DomainRange = null; // not valid
const c: DomainRange = undefined; // not valid
const d: DomainRange = {min: 1}; // valid
const e: DomainRange = {max: 2}; // valid
const f: DomainRange = { min: 3, max: 4 }; // valid
```

**Comment by sulemanof** [comment]
> We can change the `DomainRange` interface to something like this
> 
> ```ts
> // copied from stackoverflow :)
> type RequireAtLeastOne<T, Keys extends keyof T = keyof T> =
>     Pick<T, Exclude<keyof T, Keys>>
>     & {
>         [K in Keys]-?: Required<Pick<T, K>>
>     }[Keys]
> 
> type DomainRange = RequireAtLeastOne<Record<'min' | 'max', number>, 'min' | 'max'>;
> 
> const a: DomainRange = {}; // not valid
> const b: DomainRange = null; // not valid
> const c: DomainRange = undefined; // not valid
> const d: DomainRange = {min: 1}; // valid
> const e: DomainRange = {max: 2}; // valid
> const f: DomainRange = { min: 3, max: 4 }; // valid
> ```

I think in this case it wouldn't be flexible for usage and also would require additional checks before passing props..
What if change the `DomainRange` prop as an object to something like  `minDomain` and `maxDomain`?

**Comment by sulemanof** [comment]
> @sulemanof For the error throw related question: we can have two way to handle this:
> 
> * let's the developer catch that error and notify the user with the appropriate message
> * display a message instead of the chart with that error.

In this way the developer should also check the type of error and throw the appropriate. Or even check the range correctness before rendering.. Seems the question should be discussed in more detail!

**Comment by emmacunningham** [comment]
> > We can change the `DomainRange` interface to something like this
> > ```ts
> > // copied from stackoverflow :)
> > type RequireAtLeastOne<T, Keys extends keyof T = keyof T> =
> >     Pick<T, Exclude<keyof T, Keys>>
> >     & {
> >         [K in Keys]-?: Required<Pick<T, K>>
> >     }[Keys]
> > 
> > type DomainRange = RequireAtLeastOne<Record<'min' | 'max', number>, 'min' | 'max'>;
> > 
> > const a: DomainRange = {}; // not valid
> > const b: DomainRange = null; // not valid
> > const c: DomainRange = undefined; // not valid
> > const d: DomainRange = {min: 1}; // valid
> > const e: DomainRange = {max: 2}; // valid
> > const f: DomainRange = { min: 3, max: 4 }; // valid
> > ```
> 
> I think in this case it wouldn't be flexible for usage and also would require additional checks before passing props..
> What if change the `DomainRange` prop as an object to something like `minDomain` and `maxDomain`?

@markov00 @sulemanof We could also explicitly create the allowed possibilities and make `DomainRange` a union type of these possibilities.

```
export type DomainRange = LowerBoundedDomain | UpperBoundedDomain | CompleteBoundedDomain;

export interface LowerBoundedDomain {
  min: number;
}

export interface UpperBoundedDomain {
  max: number;
}

export interface CompleteBoundedDomain {
  min: number;
  max: number;
}
```

[See #116 for how this would affect the implementation]

**Comment by emmacunningham** [comment]
It seems like we're having two conversations here: how to adequately constrain the type for `DomainRange` and how to handle errors overall (in this particular case, invalid `DomainRange` configurations when `min` is greater than `max`).  To make it easier to track both discussions, I've created a separate issue #117 for the discussion on how to handle errors (not exclusive to this issue and we throw errors elsewhere in the library).

**Comment by emmacunningham** [comment]
@sulemanof Enhancement to specify only one bound for a custom domain has been released in `3.4.0` if you would like to try it out.
